src/detectors/Rust/RustComponentDetector.ts

```diff
@@ -0,0 +1,71 @@
+import { IProjectComponentDetector } from '../IProjectComponentDetector';
+import { injectable, inject } from 'inversify';
+import { Types } from '../../types';
+import { LanguageAtPath, ProjectComponent, ProjectComponentFramework, ProjectComponentPlatform, ProjectComponentType } from '../../model';
+import { IFileInspector, RustPackageInspector } from '../../inspectors';
+import lo from 'lodash';
+
+@injectable()
+export class RustComponentDetector implements IProjectComponentDetector {
+  constructor(
+    private readonly packageInspector: RustPackageInspector,
+    @inject(Types.IFileInspector) private readonly fileInspector: IFileInspector,
+  ) {}
+
+  async detectComponent(langAtPath: LanguageAtPath): Promise<ProjectComponent[]> {
+    const result: ProjectComponent[] = [];
+
+    // Cargo outputs a binary output when it finds:
+    // * `src/main.rs`
+    // * `src/bin/*.rs`
+    // * `src/bin/*/main.rs`
+    // * a `bin` entry in the manifest
+    const hasBinaryFile = Promise.all([
+      this.fileInspector.exists('src/main.rs'),
+      this.fileInspector
+        .readDirectory('src/bin')
+        .then(async (entries) => {
+          for (const entry of entries) {
+            const entryPath = `src/bin/${entry}`;
+            if (
+              (entry.endsWith('.rs') && (await this.fileInspector.isFile(entryPath))) ||
+              (await this.fileInspector.isFile(`${entryPath}/main.rs`))
+            ) {
+              return true;
+            }
+          }
+
+          return false;
+        })
+        .catch((err) => (err.code === 'ENOENT' ? Promise.resolve(false) : Promise.reject(err))),
+      Promise.resolve(lo.get(this.packageInspector.cargoManifest, 'bin.length', 0) > 0),
+    ]).then((a) => a.some((i) => i));
+
+    // Cargo outputs a library if and only if `src/lib.rs` exists
+    const hasLib = this.fileInspector.exists('src/lib.rs');
+
+    if (await hasLib) {
+      result.push({
+        framework: ProjectComponentFramework.UNKNOWN,
+        language: langAtPath.language,
+        path: langAtPath.path,
+        platform: ProjectComponentPlatform.BackEnd,
+        repositoryPath: undefined,
+        type: ProjectComponentType.Library,
+      });
+    }
+
+    if (await hasBinaryFile) {
```

Comment:
Strictly speaking you may have a standalone `Cargo.toml` file without any source files present or one with some rust sources of which neither generates any output. It is not a valid rust package, however, and Cargo will fail when attempting to build such a package. I'm not sure what to do in that case. Can you advise?

Reply:
I think it's quite the edge case that somebody would want to check a project without any source files present. We assume that people will check working projects with the dx-scanner. So my thought was to simplify code to just on if. `if (hasBinaryFile) { save the component info  - is library } else { the component is application }`. What do you think about it?

Reply:
I agree with returning something in case nothing is detected. However there is also the question of multiple return values. From the array return type I have assumed that the compoment detector may return multiple compoments for each configurations it finds.

One Cargo package can contain at most one library and 0 or more binaries, all in the same package, with the same lock file and same dependencies. This led me to believe that I should expose this by returning 2 different components from this detector. What is your opinion on that?